kernel/comps/keyboard/src/i8042_chip/mouse.rs

```diff
@@ -0,0 +1,477 @@
+            DEVICE_ID_INTELLIMOUSE | DEVICE_ID_INTELLIMOUSE_EXPLORER => {
+                Some(MouseType::IntelliMouse)
+            }
+            _ => None,
+        }
+    }
+
+    fn packet_len(&self) -> usize {
+        match self {
+            MouseType::Standard => Self::PACKET_LEN_STANDARD,
+            MouseType::IntelliMouse => Self::PACKET_LEN_INTELLIMOUSE,
+        }
+    }
+}
+
+/// Mouse packet state machine for handling 3-byte or 4-byte PS/2 mouse packets.
+struct PacketState {
+    state: PacketCollectionState,
+    packet: [u8; MouseType::PACKET_LEN_MAX],
+    byte_index: usize,
+    mouse_type: MouseType,
+    prev_left: bool,
+    prev_right: bool,
+    prev_middle: bool,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum PacketCollectionState {
+    WaitingForFirstByte,
+    CollectingPacket,
+}
+
+impl PacketState {
+    const fn new() -> Self {
+        Self {
+            state: PacketCollectionState::WaitingForFirstByte,
+            packet: [0; MouseType::PACKET_LEN_MAX],
+            byte_index: 0,
+            mouse_type: MouseType::Standard,
+            prev_left: false,
+            prev_right: false,
+            prev_middle: false,
+        }
+    }
+
+    fn set_mouse_type(&mut self, mouse_type: MouseType) {
+        self.mouse_type = mouse_type;
+        self.reset();
+    }
+
+    /// Processes a mouse data byte and returns input events if a complete packet is ready.
+    fn process_byte(&mut self, data: u8) -> Option<Vec<InputEvent>> {
+        const PACKET_SYNC_MASK: u8 = 0x08;
+
+        match self.state {
+            PacketCollectionState::WaitingForFirstByte => {
+                // First byte must have bit 3 set (always 1 in PS/2 mouse packets).
+                if data & PACKET_SYNC_MASK != 0 {
+                    self.packet[0] = data;
+                    self.byte_index = 1;
+                    self.state = PacketCollectionState::CollectingPacket;
+                }
+                None
+            }
+            PacketCollectionState::CollectingPacket => {
+                self.packet[self.byte_index] = data;
+                self.byte_index += 1;
+
+                if self.byte_index >= self.mouse_type.packet_len() {
+                    // Complete packet received.
+                    let events = self.parse_packet();
+                    self.reset();
+                    Some(events)
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /// Parses a complete PS/2 mouse packet and returns input events.
+    fn parse_packet(&mut self) -> Vec<InputEvent> {
+        const BUTTON_LEFT_MASK: u8 = 0x01;
+        const BUTTON_RIGHT_MASK: u8 = 0x02;
+        const BUTTON_MIDDLE_MASK: u8 = 0x04;
+        const INDEX_STATUS: usize = 0;
+        const INDEX_X: usize = 1;
+        const INDEX_Y: usize = 2;
+        const INDEX_WHEEL: usize = 3;
+
+        let mut events = Vec::new();
```

Comment:
To avoid multiple heap allocations, use `Vec::with_capacity()` here.
```rust
        // Currently, this method can generate at most 7 events. Don't forget to update this when
        // modifying the logic below!
        let mut events = Vec::with_capacity(7);
```